sympy/assumptions/handlers/matrices.py

```diff
@@ -6,697 +6,769 @@
-    """
-    Handler for key 'square'
-    """
-
-    @staticmethod
-    def MatrixExpr(expr, assumptions):
-        return expr.shape[0] == expr.shape[1]
-
-
-class AskSymmetricHandler(CommonHandler):
-    """
-    Handler for key 'symmetric'
-    """
-
-    @staticmethod
-    def MatMul(expr, assumptions):
-        factor, mmul = expr.as_coeff_mmul()
-        if all(ask(Q.symmetric(arg), assumptions) for arg in mmul.args):
-            return True
-        # TODO: implement sathandlers system for the matrices.
-        # Now it duplicates the general fact: Implies(Q.diagonal, Q.symmetric).
-        if ask(Q.diagonal(expr), assumptions):
-            return True
-        if len(mmul.args) >= 2 and mmul.args[0] == mmul.args[-1].T:
-            if len(mmul.args) == 2:
-                return True
-            return ask(Q.symmetric(MatMul(*mmul.args[1:-1])), assumptions)
-
-    @staticmethod
-    def MatPow(expr, assumptions):
-        # only for integer powers
-        base, exp = expr.args
-        int_exp = ask(Q.integer(exp), assumptions)
-        if not int_exp:
-            return None
-        non_negative = ask(~Q.negative(exp), assumptions)
-        if (non_negative or non_negative == False
-                            and ask(Q.invertible(base), assumptions)):
-            return ask(Q.symmetric(base), assumptions)
-        return None
+def BM_elements(predicate, expr, assumptions):
+    """ Block Matrix elements """
+    return all(ask(predicate(b), assumptions) for b in expr.blocks)
 
-    @staticmethod
-    def MatAdd(expr, assumptions):
-        return all(ask(Q.symmetric(arg), assumptions) for arg in expr.args)
-
-    @staticmethod
-    def MatrixSymbol(expr, assumptions):
-        if not expr.is_square:
-            return False
-        # TODO: implement sathandlers system for the matrices.
-        # Now it duplicates the general fact: Implies(Q.diagonal, Q.symmetric).
-        if ask(Q.diagonal(expr), assumptions):
-            return True
-        if Q.symmetric(expr) in conjuncts(assumptions):
-            return True
+def MS_elements(predicate, expr, assumptions):
+    """ Matrix Slice elements """
+    return ask(predicate(expr.parent), assumptions)
 
-    @staticmethod
-    def ZeroMatrix(expr, assumptions):
-        return ask(Q.square(expr), assumptions)
+def MatMul_elements(matrix_predicate, scalar_predicate, expr, assumptions):
+    d = sift(expr.args, lambda x: isinstance(x, MatrixExpr))
+    factors, matrices = d[False], d[True]
+    return fuzzy_and([
+        test_closed_group(Basic(*factors), assumptions, scalar_predicate),
+        test_closed_group(Basic(*matrices), assumptions, matrix_predicate)])
 
-    OneMatrix = ZeroMatrix
 
-    @staticmethod
-    def Transpose(expr, assumptions):
-        return ask(Q.symmetric(expr.arg), assumptions)
+### AskSquareHandler ###
 
-    Inverse = Transpose
+AskSquareHandler = CommonHandler.copy(
+    'AskSquareHandler',
+    doc="""
+    Handler for key 'square'
+    """
+)
```

Comment:
I don't understand why we need all these objects - why can't we use `Q.square.handler.register` instead of `AskSquareHandler.register`? Is it an import issue? A lifetime issue?

Reply:
We can use `Q.square.handler.register` if `AskSquareHandler` is registered to `Q.square`. I updated the document - please see `ask.register_handler`.

I implemented `copy` method because `AskSquareHandler` must inherit every function dispatched to `CommonHandler`. Previously, these two were classes and inheritance was done by subclassing. But in this design, these two are multipledispatch instance so I had to make this change.

If you have any other idea please let me know. Thanks!

Reply:
What handlers does `CommonHandler` have that we actually care about? Are there any cases where there are multiple levels of base classes, `SomeHandler -> Other handler -> CommonHandler`? Or is it always just one level?

Reply:
In  `assumptions.handlers.sets`, we have `AskExtendedRealHandler`  which is derived from `AskRealHandler`. It's `AskExtendedRealHAndler -> AskRealHandler -> CommonHandler`.

Reply:
Could this be handled by registering a dispatcher default that calls the "base" handler?

Reply:
Perhaps I can add `fallback_dispatcher` attribute to `AskHandlerClass`. I will try that way and report here.

Reply:
I removed `copy` method and added `base` parameter to `AskHandlerClass.__init__` in commit `f3247ab`. Now, the syntax is
```python
AskSquareHandler = HandlerClass('AskSquareHandler', doc='...', base=CommonHandler)
```
Here, `AskSquareHandler` inherits every functions dispatched to `CommonHandler`.

Reply:
@eric-wieser, if you have any other comment, please leave your review in #20475. Thanks!